### ISSUE TITLE
Permission denied (publickey).

### ISSUE THREAD

**Issue by briankitt** [sshnet/SSH.NET#273]
I am trying to connect with a server using SSH.  I am at a loss, been struggling with this for a few days, and I cannot get this to work.  I am connecting to USBank, and they do this routinely on a daily basis, so I assume they are not making a mistake, the mistake has to be on my end, but I don't get what I am doing wrong.

I keep getting the error in the subject line.  I have generated SSh keys with Git Bash (sh-keygen -t rsa -b 2048), and also with openssl (openssl genrsa -des3 -out private.pem 2048     openssl rsa -in private.pem -outform PEM -pubout -out public.pem).  Both keys give me the same error.
Here is my code:
PrivateKeyFile[] keyFiles = new PrivateKeyFile[1];
keyFiles[0] = new PrivateKeyFile(<<filename>>, ConfigurationManager.AppSettings["RSAKey"]);
// note, filename is the private key that was generated from Git
//-----BEGIN RSA PRIVATE KEY-----
//Proc-Type: 4,ENCRYPTED
//DEK-Info: DES-EDE3-CBC,46A6437F2613E570
AuthenticationMethod[] authenticationMethods = new AuthenticationMethod[1];
authenticationMethods[0] = new PrivateKeyAuthenticationMethod(<<UserName>>, keyFiles)
ConnectionInfo connectionInfo= new ConnectionInfo(<<Hos>>, <<Port>>, <<UserName>>, authenticationMethods);
using (SftpClient client = new SftpClient(connectionInfo))
{
       client.Connect();
       IEnumerable<SftpFile> files = client.ListDirectory("/Inbox");
}

The public key that I give the client starts with:
-----BEGIN PUBLIC KEY-----

**Comment by darinkes** [comment]
Does it work, if you use your private key with openssh client?
ssh -i id_rsa user@host

maybe adding some -vvvv to the client might help to enlight the situation.

Edit: Please use the Code-Tags to make it readable for us :)

**Comment by briankitt** [comment]
I don't have access to a server where I could place my public key, so I
can't test it at all.
How would I add -vvvv to the C# API?

**Comment by darinkes** [comment]
Its hard to follow the explanations. The Public Key is on the server or not?

If it is, test it with openssh client.
If it isn't, it won't work with any ssh client/library.

**Comment by briankitt** [comment]
Not sure what is hard to understand here.

I generated a public/private key pair using Git Bash.
I gave the public key to US Bank.
I am using the private key to connect to them using this API.
I get the error in my subject line.  I couldn't get that to work.

I then generated a public/private key pair using open SSL (in case Git Bash
was in error).
I gave the public key to US Bank.
Again, I am using the private key to connect to them using this API.
I still get the error.

US Bank creates connections like this on a daily basis, I can only assume
they are doing everything correctly.

The problem has to be in this API.  Why can't I connect?  Why am I getting
this error?

**Comment by darkoperator** [comment]
To be honest best way to test would be to setup a test Linux VM and configure the key on it to test.

**Comment by briankitt** [comment]
I don't think anyone is understanding my issue.  I don't have time to
banter back and forth.

What other products are out there that can do SSH file transfer.  This one
doesn't work.

**Comment by darinkes** [comment]
Ok. So the key is on the server. That was not clear from your writings.
Then open git bash, where you generated the key, and run: ssh -vvvv -i your-private-key user@host.

**Comment by darkoperator** [comment]
We do but without proper testing we can not tell you what you are doing wrong. It is not not the library. My powershelgl module Posh-SSH gets hundreds of download a day and it is used constantly and it has worked for all those people. You can use some of the code here as example https://github.com/darkoperator/Posh-SSH/tree/master/PoshSSH <https://github.com/darkoperator/Posh-SSH/tree/master/PoshSSH>

**Comment by briankitt** [comment]
Thanks!  I see that you simply wanted to test the key, I thought you were
asking me to change my program.  I get that now.

It doesn't connect.  WY.private.asc is the file name of my private file.

So this is telling me that USBank messed up, three different times?

OpenSSH_7.3p1, OpenSSL 1.0.2j  26 Sep 2016
debug1: Reading configuration data /etc/ssh/ssh_config
debug2: resolving <removed for security>
debug2: ssh_connect_direct: needpriv 0
debug1: Connecting to <removed for security>
debug1: Connection established.
debug1: key_load_public: No such file or directory
debug1: identity file WY.private.asc type -1
debug1: key_load_public: No such file or directory
debug1: identity file WY.private.asc-cert type -1
debug1: Enabling compatibility mode for protocol 2.0
debug1: Local version string SSH-2.0-OpenSSH_7.3
debug1: Remote protocol version 2.0, remote software version Maverick_SSHD
debug1: no match: Maverick_SSHD
debug2: fd 3 setting O_NONBLOCK
debug1: Authenticating to <removed for security>
debug3: put_host_port: <removed for security>
debug3: send packet: type 20
debug1: SSH2_MSG_KEXINIT sent
debug3: receive packet: type 20
debug1: SSH2_MSG_KEXINIT received
debug2: local client KEXINIT proposal
debug2: KEX algorithms: curve25519-sha256@libssh.org
,ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,diffie-hellman-group-exchange-sha256,diffie-hellman-group16-sha512,diffie-hellman-group18-sha512,diffie-hellman-group-exchange-sha1,diffie-hellman-group14-sha256,diffie-hellman-group14-sha1,ext-info-c
debug2: host key algorithms: ecdsa-sha2-nistp256-cert-v01@openssh.com,
ecdsa-sha2-nistp384-cert-v01@openssh.com,
ecdsa-sha2-nistp521-cert-v01@openssh.com,ssh-ed25519-cert-v01@openssh.com,
ssh-rsa-cert-v01@openssh.com
,ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,ssh-ed25519,rsa-sha2-512,rsa-sha2-256,ssh-rsa
debug2: ciphers ctos: chacha20-poly1305@openssh.com
,aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,
aes256-gcm@openssh.com,aes128-cbc,aes192-cbc,aes256-cbc,3des-cbc
debug2: ciphers stoc: chacha20-poly1305@openssh.com
,aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,
aes256-gcm@openssh.com,aes128-cbc,aes192-cbc,aes256-cbc,3des-cbc
debug2: MACs ctos: umac-64-etm@openssh.com,umac-128-etm@openssh.com,
hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,
hmac-sha1-etm@openssh.com,umac-64@openssh.com,umac-128@openssh.com
,hmac-sha2-256,hmac-sha2-512,hmac-sha1
debug2: MACs stoc: umac-64-etm@openssh.com,umac-128-etm@openssh.com,
hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,
hmac-sha1-etm@openssh.com,umac-64@openssh.com,umac-128@openssh.com
,hmac-sha2-256,hmac-sha2-512,hmac-sha1
debug2: compression ctos: none,zlib@openssh.com,zlib
debug2: compression stoc: none,zlib@openssh.com,zlib
debug2: languages ctos:
debug2: languages stoc:
debug2: first_kex_follows 0
debug2: reserved 0
debug2: peer server KEXINIT proposal
debug2: KEX algorithms:
diffie-hellman-group1-sha1,diffie-hellman-group-exchange-sha256,diffie-hellman-group-exchange-sha1
debug2: host key algorithms: ssh-rsa
debug2: ciphers ctos:
aes128-ctr,aes256-ctr,aes192-ctr,aes256-cbc,aes192-cbc,3des-cbc
debug2: ciphers stoc:
aes128-ctr,aes256-ctr,aes192-ctr,aes256-cbc,aes192-cbc,3des-cbc
debug2: MACs ctos:
hmac-sha2-512,hmac-sha2-256,hmac-sha2-512-96,hmac-sha2-256-96,hmac-sha1,hmac-sha1-96
debug2: MACs stoc:
hmac-sha2-512,hmac-sha2-256,hmac-sha2-512-96,hmac-sha2-256-96,hmac-sha1,hmac-sha1-96
debug2: compression ctos: none
debug2: compression stoc: none
debug2: languages ctos:
debug2: languages stoc:
debug2: first_kex_follows 0
debug2: reserved 0
debug1: kex: algorithm: diffie-hellman-group-exchange-sha256
debug1: kex: host key algorithm: ssh-rsa
debug1: kex: server->client cipher: aes128-ctr MAC: hmac-sha2-256
compression: none
debug1: kex: client->server cipher: aes128-ctr MAC: hmac-sha2-256
compression: none
debug3: send packet: type 34
debug1: SSH2_MSG_KEX_DH_GEX_REQUEST(2048<8192<8192) sent
debug3: receive packet: type 31
debug1: got SSH2_MSG_KEX_DH_GEX_GROUP
debug2: bits set: 985/2048
debug3: send packet: type 32
debug1: SSH2_MSG_KEX_DH_GEX_INIT sent
debug3: receive packet: type 33
debug1: got SSH2_MSG_KEX_DH_GEX_REPLY
debug1: Server host key: ssh-rsa
SHA256:4xGJ9WvBzEymJDnJ5JVylMT9xDtfVEdYelmxhWBdLqA
debug3: put_host_port: <removed for security>
debug3: put_host_port: <removed for security>
debug1: checking without port identifier
The authenticity of host '<removed for security>' can't be established.
RSA key fingerprint is SHA256:<removed for security>
Are you sure you want to continue connecting (yes/no)? yes
Warning: Permanently added '<removed for security>' (RSA) to the list of
known hosts.
debug2: bits set: 1027/2048
debug3: send packet: type 21
debug2: set_newkeys: mode 1
debug1: rekey after 4294967296 blocks
debug1: SSH2_MSG_NEWKEYS sent
debug1: expecting SSH2_MSG_NEWKEYS
debug3: receive packet: type 21
debug2: set_newkeys: mode 0
debug1: rekey after 4294967296 blocks
debug1: SSH2_MSG_NEWKEYS received
debug2: key: WY.private.asc (0x0), explicit
debug3: send packet: type 5
debug3: receive packet: type 6
debug2: service_accept: ssh-userauth
debug1: SSH2_MSG_SERVICE_ACCEPT received
debug3: send packet: type 50
debug3: receive packet: type 53
debug3: input_userauth_banner
UAT SFTP Testing  on 6022 in SSP ENgine 1
debug3: receive packet: type 51
debug1: Authentications that can continue:
password,publickey,keyboard-interactive
debug3: start over, passed a different list
password,publickey,keyboard-interactive
debug3: preferred publickey,keyboard-interactive,password
debug3: authmethod_lookup publickey
debug3: remaining preferred: keyboard-interactive,password
debug3: authmethod_is_enabled publickey
debug1: Next authentication method: publickey
debug1: Trying private key: WY.private.asc
Enter passphrase for key 'WY.private.asc':
debug3: sign_and_send_pubkey: RSA
SHA256:ObYR/iiaSdvAdMCoSLOWnldEg1cLlAshpW/YZkrjpM0
debug3: send packet: type 50
debug2: we sent a publickey packet, wait for reply
debug3: receive packet: type 51
debug1: Authentications that can continue:
password,publickey,keyboard-interactive
debug2: we did not send a packet, disable method
debug3: authmethod_lookup keyboard-interactive
debug3: remaining preferred: password
debug3: authmethod_is_enabled keyboard-interactive
debug1: Next authentication method: keyboard-interactive
debug2: userauth_kbdint
debug3: send packet: type 50
debug2: we sent a keyboard-interactive packet, wait for reply
debug3: receive packet: type 60
debug2: input_userauth_info_req
password
Enter Password for <removed for security>:

**Comment by briankitt** [comment]
Thanks, but I've tried dozens of samples of sample code.  None of them
work.  Not really looking for any more sample code.

The guy at USBank tells me he did everything right.  I've done this before
with other banks.  I don't get why it's not working.  The problem is, that
I have virtually nothing do go on, other then I am getting a message
indicating a key mismatch.  Logic tells me that the guy at USBank had to
mess up, but he tells me it is correct.  My hands are really tied.  I'm
getting pressure on me to resolve this, everyone is pointing fingers at me,
but I don't understand why it's not working.  I've used 2 different
utilities to generate RSA keys, I've generated 3 sets of keys.  None of
them are working, I keep getting the same error.

I'm at a loss of where to go.  I don't have access to a server that I can
set up an SSH site, so I am really handicapped on this one.

**Comment by darinkes** [comment]
debug3: sign_and_send_pubkey: RSA
SHA256:ObYR/iiaSdvAdMCoSLOWnldEg1cLlAshpW/YZkrjpM0
debug3: send packet: type 50
debug2: we sent a publickey packet, wait for reply
debug3: receive packet: type 51

Means the authentication was declined by the server. And if openssh fails, the failure must be on the server side.

Maybe the private key got corrupted during transfer to the bank? Encoding, blanks, seen it all...

**Comment by briankitt** [comment]
Thanks!

I've been trying to tell them that, and they keep denying that.  Everyone
is pointing fingers at me, telling me that my code is wrong (this API).

I'll take this debug information, and send it back to USBank.

Thanks so much for your help, this is hugely appreciated.  I've been
extremely frustrated that I cannot get any debug information.

**Comment by darkoperator** [comment]
I think it is because you uses OpenSSL, switch to the ssh-keygen instead https://security.stackexchange.com/questions/32768/converting-keys-between-openssl-and-openssh <https://security.stackexchange.com/questions/32768/converting-keys-between-openssl-and-openssh> That I know off SSH.Net <http://ssh.net/> does not work with OpenSSL PEM format keys.

**Comment by briankitt** [comment]
BINGO!

That fixed it!   Now it works fine!

Thanks so much!

**Comment by Ben555555** [comment]
In our case the reason for this error message was that the username is case sensitive.
